File: gnu/gcc/libgomp/testsuite/libgomp.fortran/nestedfn1.f90
! { dg-do run }

  integer :: a, b, c
  a = 1
  b = 2
  c = 3
  call foo
  if (a .ne. 7) call abort
contains
  subroutine foo
    use omp_lib
    logical :: l
    l = .false.
!$omp parallel shared (a) private (b) firstprivate (c) &
!$omp num_threads (2) reduction (.or.:l)
    if (a .ne. 1 .or. c .ne. 3) l = .true.
!$omp barrier
    if (omp_get_thread_num () .eq. 0) then
      a = 4
      b = 5
      c = 6
    end if
!$omp barrier
    if (omp_get_thread_num () .eq. 1) then
      if (a .ne. 4 .or. c .ne. 3) l = .true.
      a = 7
      b = 8
      c = 9
    else if (omp_get_num_threads () .eq. 1) then
      a = 7
    end if
!$omp barrier
    if (omp_get_thread_num () .eq. 0) then
      if (a .ne. 7 .or. b .ne. 5 .or. c .ne. 6) l = .true.
    end if
!$omp barrier
    if (omp_get_thread_num () .eq. 1) then
      if (a .ne. 7 .or. b .ne. 8 .or. c .ne. 9) l = .true.
    end if
!$omp end parallel
    if (l) call abort
  end subroutine foo
end

File: gnu/gcc/libgomp/testsuite/libgomp.c/nestedfn-1.c
/* { dg-do run } */

#include <omp.h>
#include <stdlib.h>

int
main (void)
{
  int a = 1, b = 2, c = 3;
  void
  foo (void)
  {
    int l = 0;
#pragma omp parallel shared (a) private (b) firstprivate (c) \
		     num_threads (2) reduction (||:l)
    {
      if (a != 1 || c != 3) l = 1;
#pragma omp barrier
      if (omp_get_thread_num () == 0)
	{
	  a = 4;
	  b = 5;
	  c = 6;
	}
#pragma omp barrier
      if (omp_get_thread_num () == 1)
	{
	  if (a != 4 || c != 3) l = 1;
	  a = 7;
	  b = 8;
	  c = 9;
	}
      else if (omp_get_num_threads () == 1)
	a = 7;
#pragma omp barrier
      if (omp_get_thread_num () == 0)
	if (a != 7 || b != 5 || c != 6) l = 1;
#pragma omp barrier
      if (omp_get_thread_num () == 1)
	if (a != 7 || b != 8 || c != 9) l = 1;
    }
    if (l)
      abort ();
  }
  foo ();
  if (a != 7)
    abort ();
  return 0;
}
